drivers/wdm/audio/backpln/portcls/irpstream.cpp

```diff
@@ -779,23 +780,50 @@ CIrpQueue::GetCurrentIrpOffset()
 BOOLEAN
 NTAPI
 CIrpQueue::GetAcquiredTagRange(
-    IN PVOID * FirstTag,
-    IN PVOID * LastTag)
+    OUT PVOID * FirstTag,
+    OUT PVOID * LastTag)
 {
     KIRQL OldLevel;
     BOOLEAN Ret = FALSE;
-    //PIRP Irp;
-    //PLIST_ENTRY CurEntry;
-    //PKSSTREAM_DATA StreamData;
+    PIRP Irp;
+    PLIST_ENTRY CurEntry;
+    PKSSTREAM_DATA StreamData;
 
     // lock list
     KeAcquireSpinLock(&m_IrpListLock, &OldLevel);
 
-    // initialize to zero
-    *FirstTag = NULL;
-    *LastTag = NULL;
+    CurEntry = m_IrpList.Flink;
+
+    while (CurEntry != &m_IrpList)
+    {
+        // get irp from list entry
+        Irp = (PIRP)CONTAINING_RECORD(CurEntry, IRP, Tail.Overlay.ListEntry);
+
+        // get stream data
+        StreamData = (PKSSTREAM_DATA)Irp->Tail.Overlay.DriverContext[STREAM_DATA_OFFSET];
+
+        if (StreamData->StreamHeaderIndex == 0)
+        {
+            // store the first tag from stream header index
+            *FirstTag = StreamData->Tags[StreamData->StreamHeaderIndex].Tag;
+        }
 
-    UNIMPLEMENTED;
+        // increment stream header index
+        StreamData->StreamHeaderIndex++;
```

Comment:
It doesn't seem like you should modify this. It seems to indicate a "current position" of some type, whereas this operation seems like it should be read-only. This is based on a cursory reading, so I may be wrong.